### PR TITLE
Ensures spring context is closed in brave tests

### PR DESCRIPTION
**Pull request by adriancole**
This also normalizes how we subscribe to request monos.

### CODE REVIEW

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/HttpClientBeanPostProcessor.java [hunk 1]
```diff
@@ -182,6 +182,7 @@ static void parseConnectionAddress(Connection connection, Span span) {
 
 		@Override
 		public void accept(HttpClientResponse response, Connection connection) {
+			// TODO: is there a way to read the request at response time?
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/HttpClientBeanPostProcessor.java]: @violetagg is there any way to do this at the moment or some future release of reactor-netty? The intent is to see the request that actually caused the response. The original request may have been rewritten due to redirects etc.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java [hunk 2]
```diff
@@ -274,6 +276,7 @@ public void onNext(ClientResponse response) {
 								.build());
 			}
 			finally {
+				// TODO: is there a way to read the request at response time?
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java]: @rstoyanchev @bclozel  is there any way to do this at the moment or some future release of reactor-netty? The intent is to see the request that actually caused the response. The original request may have been rewritten due to redirects etc.

Comment (adriancole) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java]: updated the copy/paste :P

File: tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java [hunk 3]
```diff
@@ -107,14 +99,12 @@ public void onSubscribe(Subscription s) {
 			}
 
 			@Override
-			public void onNext(ClientResponse t) {
-				Subscription s = ref.getAndSet(null);
-				if (s != null) {
+			public void onNext(ClientResponse clientResponse) {
+				if (ref.getAndSet(null) != null) {
 					callback.onSuccess(null);
-					s.cancel();
```
Comment (adriancole) [inline tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java]: the other test doesn't call cancel, but I noticed sometimes this is called. This causes a problem really, if we assume that cancel is an error. For example, a success response might not happen due to this cancel depending on order cc @simonbasle @robotmrv

File: tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java [hunk 4]
```diff
@@ -128,7 +118,8 @@ public void onError(Throwable t) {
 			@Override
 			public void onComplete() {
 				if (ref.getAndSet(null) != null) {
-					callback.onSuccess(null);
+					callback.onError(
```
Comment (simonbasle) [inline tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java]: isn't that path testing the "empty source" case? ie. if the `ref` wasn't set to `null` previously, it means that we completed directly, aka. the source is empty. this is not an error. note that a well-formed source MUST guarantee `onNext`-before-`onComplete` order.

File: tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java (hunk 3 above)
Comment (simonbasle) [reply tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java]: `s` is the object that produced the `ClientResponse`, the `Subscription` generated by the `request` Mono, and if we're in `onNext` it means that it already produced it... so `s.cancel()` doesn't bring much _for that particular case_.

if the test could run on an arbitrary source, then `s.cancel()` would make sense to signal we're only interested in the first incoming element.

File: tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java [hunk 5]
```diff
@@ -107,14 +99,12 @@ public void onSubscribe(Subscription s) {
...
-				Subscription s = ref.getAndSet(null);
-				if (s != null) {
+			public void onNext(ClientResponse clientResponse) {
+				if (ref.getAndSet(null) != null) {
 					callback.onSuccess(null);
-					s.cancel();
 				}
 				else {
-					Operators.onNextDropped(t, currentContext());
+					Operators.onNextDropped(clientResponse, currentContext());
```
Comment (simonbasle) [inline tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java]: again, with a `Mono` source this should never happen... maybe make it fail the test rather than `onNextDropped` (which by default only logs, IIRC).

File: tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java (hunk 4 above)
Comment (simonbasle) [reply tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java]: again, since that is dedicated to a specific test, having such an early completion could indeed be considered an error (ie. we expect exactly `onNext(ClientResponse)` then `onComplete()`).

Comment (adriancole) [reply tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java]: we have so few tests about these edge cases.. https://github.com/spring-cloud/spring-cloud-sleuth/issues/1570

main thing is that the two types of integration tests here, WebClient and HttpClient should have a coherent means to do the subscription dance, or a clear reason why they are different. This PR so far attempts to get them to do the same thing.. then we can start documenting all the odd possibilities therein

File: tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java (hunk 5 above)
Comment (adriancole) [reply tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java]: sounds good

**Comment by adriancole** [comment]
I'll do my best to parse the existing mentions as well the one I've not gotten to (#1570) tomorrow. appreciate the responses so far and keep them coming. Unlike webflux, I'm ok with multiple responses prior to onComplete.

File: tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java (hunk 3 above)
Comment (adriancole) [reply tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java]: thanks.. moved the thing about arbitrary source here https://github.com/spring-cloud/spring-cloud-sleuth/issues/1570#issuecomment-591247178

File: tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java (hunk 4 above)
Comment (adriancole) [reply tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/WebClientBraveTests.java]: empty source, we'll handle explicitly here. https://github.com/spring-cloud/spring-cloud-sleuth/issues/1570#issuecomment-591247178

thanks for the reminder. this particular one is test code which isn't written to also handle empty source, plus there's no empty source test, yet

File: tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/ReactorNettyHttpClientBraveTests.java [hunk 6]
```diff
@@ -82,13 +66,15 @@ static HttpClient testHttpClient() {
 	}
 
 	@Override
-	protected void closeClient(HttpClient client) {
-		// HttpClient is not Closeable
+	protected void closeClient(AnnotationConfigApplicationContext context) {
+		context.close(); // ensures shutdown hooks fire
```
Comment (adriancole) [inline tests/spring-cloud-sleuth-instrumentation-reactor-tests/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/ReactorNettyHttpClientBraveTests.java]: reactor hooks are a PreDestroy thing.. we formerly needed hacks because this test leaked spring contexts.

**Comment by adriancole** [comment]
any more comments after are welcome

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java (hunk 2 above)
Comment (rstoyanchev) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java]: That would be an `ExchangeFilterFunction` which this [processor is using](https://github.com/spring-cloud/spring-cloud-sleuth/blob/0227c050985ca656306e50eb8ab0c69792b89420/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java#L139-L149). What you do from there, extending `Mono<ClientResponse>` and wrapping its `Subscriber` is not an expected way to use the API. Nevertheless the request seems to be passed into the former but not into the latter, but probably could be.

Comment (adriancole) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java]: thanks for the hint! added it to the todo bucket :) https://github.com/spring-cloud/spring-cloud-sleuth/pull/1573/files#diff-bb63d22a058c70307dc3b4e017e8b0b4R98
